modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js

```diff
@@ -20,74 +20,33 @@ import {Editor} from './Editor';
 import '../css/main.scss';
 
 const BalloonEditor = ({config = {}, contents, name, ...otherProps}) => {
-	const defaultExtraPlugins = 'balloontoolbar,floatingspace';
-
-	const getConfig = () => {
-		const extraPlugins = config.extraPlugins
-			? `${config.extraPlugins}`
-			: '';
-
-		return {
-			...config,
-			extraAllowedContent: '*',
-			extraPlugins: `${extraPlugins}${defaultExtraPlugins}`,
-		};
-	};
-
 	return (
 		<Editor
-			config={getConfig()}
+			config={config}
```

Comment:
@ambrinchaudhary I'm still not sure about this approach so it would be nice to clarify if we want to be able to use this component without Java ... but maybe there's something I don't understand

For example: What if you want your React component to render `BalloonEditor`?
Does that mean you have to "glue" something in between to get the config from the config contributor?

Reply:
Yes, that is what I meant [here](https://github.com/liferay-frontend/liferay-portal/pull/960#issuecomment-818595271). You need to get the config via Java, and then pass it to your react component. Lima is already doing it, and I think DE too. This way, you still have one single point to configure your editor (the _*ConfigContributors.java_), and the editor could be rendered via taglib or via JS.

Reply:
Thanks for making that clear.

Now I have another question, is this what **we** really want?

You said that Lima is doing it and DE are doing it, but that's doesn't answer my question: as a (**frontend**) infrastructure team, I think that we should take this decision based on what we think is going to be easier/better for other teams (and more specifically for the frontend developers working in other teams because they are the ones who are going to use this component) and not the other way around. At the end of the day, each team uses our infrastructure they way they want to use it (this is not a negative comment, it's just what I've observed and it makes sense since every team has their use cases), but I really think that the important point here is that the easier we make it the happier they will be (and so will we, because we have to maintain this).

So here's another take on my question: as a frontend developer, do you think depending on Java to get a the configuration for React component is ideal?

I personally think that would be a **bad idea**, because it makes things more complex.
Maybe you're used to reading Java code but I think most **frontend** developers aren't (and I'm including myself). 

Of course, if  the technical requirements specifies that the configuration must be written in Java (I'll have a look too) we can do it, but I just think that's just adding unnecessary complexity.

Reply:
I see @julien, but my point of view here is different. Until now editors where rendered via taglib, and it was not like related to frontend people, but also backends could configure it. Current java configuration allows you easily to have different editor configurations for a custom portlet (for example in wiki you have `html` editor config and `creole` config), or differents editors in the same portlet (like blogs, where you have 3 or 4 editors at the same time, each of them different). And we can still take advantage of this current way of configuration for rendering editors via JS. If we now want to think in another way of customization for editor's configuration from the frontend point of view, we will need time to think about it, and we are adding an extra way to do that just for frontend developers (because I guess we should maintain the current way)... so... extra code to maintain. 

In conclusion, in the teams I think we are more than used to this configuration way, and... I'm not sure this is something related exclusively to frontend infrastructure team.... that is only my personal point of view, I may be wrong of course... 

So let's see if someone else have more things to add...  /cc @liferay-frontend

Reply:
> Current java configuration allows you easily to have different editor configurations for a custom portlet 

You're right, but my take on it is that the Java configuration is just "copying" what CKEditor provides, at the end it's just an object with options and you can do exactly the same thing with JS.

So I agree with you that we still need to be able to configure the editor via Java, but we also need to support configuring an editor only with JS.

> and we are adding an extra way to do that just for frontend developers

That would make sense for me, since we're frontend infrastructure and not backend infrastructure.Sorry to repeat myself, but the way I see it is that we added an *extra* way to be able to use them via Java, remember: CKEditor is a JS library not a Java library.

> In conclusion, in the teams I think we are more than used to this configuration way, and

That's what you think, and it's a valid opinion, but I doubt every frontend developer thinks the same (a with me you already have an example :joy: ), so why not support both ways ?

Reply:
Additionally, one thing that I don't understand is [this](https://github.com/liferay-frontend/liferay-portal/pull/960/files#diff-91a08522aea78564ab2061bea292615577dc05c3db94c579dd042dd8c5107544R62-R63), and to save a click I'll put it here

```java
String balloonToolbarPlugins =
	"balloontoolbar,floatingspace,stylescombo";

String extraPlugins = jsonObject.getString("extraPlugins");

if (Validator.isNotNull(extraPlugins)) {
	extraPlugins += "," + balloonToolbarPlugins;
}
else {
	extraPlugins = balloonToolbarPlugins;
}
```

If I understand correctly, `EditorConfigContributor`'s should 

> Provides an interface for setting the editor's configuration

(*Found [here](https://docs.liferay.com/portal/7.0/javadocs/portal-kernel/com/liferay/portal/kernel/editor/configuration/EditorConfigContributor.html), which might be "old" but it seems like a valid definition*)

And in this case, what puzzles me is that without the `balloontoolbar` (and `floatingspace`) plugins we're not going to have a "balloon editor" but a "classic editor", so I think that those two plugins should at least be in the JS (or at least always present in the editor's configuration), to make sure that no matter what "options" are provided by the `EditorConfigContributor` class we always get a "balloon editor" ... and a balloon editor without toolbars doesn't do much, so I also think we need a basic set of toolbars defined somewhere too.

I know this is a long conversation we're having here, but I'm just trying to make sure we don't miss anything.

Reply:
As discussed in our meeting with Daniel Sanz, there will be some basic config in the Balloon Editor that could be completed or overrided (in the toolbar cases) with the ConfigContributors. See https://github.com/liferay-frontend/liferay-portal/pull/984/commits/71bee6799b2b07e146d73249482957456b2a6a8a